pkg/client/record/events_cache.go

```diff
@@ -17,85 +17,331 @@ limitations under the License.
+	// record our new observation
+	e.cache.Add(
+		key,
+		eventLog{
+			count:           event.Count,
+			firstTimestamp:  event.FirstTimestamp,
+			name:            event.Name,
+			resourceVersion: event.ResourceVersion,
+		},
+	)
 }
 
-func (eventCache *historyCache) getEventFromCache(key string) history {
-	value, ok := eventCache.cache.Get(key)
+// lastEventObservationFromCache returns the event from the cache, reads must be protected via external lock
+func (e *eventLogger) lastEventObservationFromCache(key string) eventLog {
+	value, ok := e.cache.Get(key)
 	if ok {
-		historyValue, ok := value.(history)
+		observationValue, ok := value.(eventLog)
 		if ok {
-			return historyValue
+			return observationValue
 		}
 	}
-	return history{}
+	return eventLog{}
 }
 
-func getEventKey(event *api.Event) string {
-	return event.Source.Component +
-		event.Source.Host +
-		event.InvolvedObject.Kind +
-		event.InvolvedObject.Namespace +
-		event.InvolvedObject.Name +
-		string(event.InvolvedObject.UID) +
-		event.InvolvedObject.APIVersion +
-		event.Reason +
-		event.Message
+// EventCorrelator processes all incoming events and performs analysis to avoid overwhelming the system.  It can filter all
+// incoming events to see if the event should be filtered from further processing.  It can aggregate similar events that occur
+// frequently to protect the system from spamming events that are difficult for users to distinguish.  It performs de-duplication
+// to ensure events that are observed multiple times are compacted into a single event with increasing counts.
+type EventCorrelator struct {
+	// the function to filter the event
+	filterFunc EventFilterFunc
+	// the object that performs event aggregation
+	aggregator *EventAggregator
+	// the object that observes events as they come through
+	logger *eventLogger
+}
+
+// EventCorrelateResult is the result of a Correlate
+type EventCorrelateResult struct {
+	// the event after correlation
+	Event *api.Event
+	// if provided, perform a strategic patch when updating the record on the server
+	Patch []byte
+	// if true, do no further processing of the event
+	Skip bool
+}
+
+// NewEventCorrelator creates an object with default values for correlation by reason
+// TODO come back and let the aggregate max and aggregate interval be tuned
+func NewEventCorrelator(clock util.Clock) *EventCorrelator {
+	cacheSize := maxLruCacheEntries
+	return &EventCorrelator{
+		filterFunc: DefaultEventFilterFunc,
+		aggregator: NewEventAggregator(
+			cacheSize,
+			EventAggregatorByReasonFunc,
+			EventAggregatorByReasonMessageFunc,
+			defaultAggregateMaxEvents,
+			defaultAggregateIntervalInSeconds,
+			clock),
+		logger: newEventLogger(cacheSize, clock),
+	}
+}
+
+// EventCorrelate filters, aggregates, counts, and de-duplicates all incoming events
+func (c *EventCorrelator) EventCorrelate(newEvent *api.Event) (*EventCorrelateResult, error) {
+	if c.filterFunc(newEvent) {
+		return &EventCorrelateResult{Skip: true}, nil
```

Comment:
It was a placeholder for if/when I wanted to add the ability to just ignore too-frequent events and drop them client-side.